### ISSUE TITLE
【AIコース】大課題 ボストン住宅価格予測 決定木

### ISSUE THREAD

**Issue by hayyoshi** [hayyoshi/DIC#3]
### データを取得する
from sklearn.datasets import load_boston
boston = load_boston()

### 取得データをDataFrameにする
import pandas as pd
# columnsにはload_boston().feature_namesを指定して、カラム名を指定します。
# もしくはインスタンス.feature_names
X = pd.DataFrame(boston.data,columns=boston.feature_names)
# columnsには、'MEDV'を指定します。
y = pd.DataFrame(boston.target,columns=['MEDV'])

### 基本的なデータ探索について記述せよ
# このデータセットは何のデータセットであるか(何のために作成されたか、誰が作成したか、何年に作成されたか)
# 米国ボストン市郊外の地域別の13種類の特徴と住宅価格の統計情報です。
# 住宅価格の線形回帰分析を目的として、Harrison, D. and Rubinfeld, D.L.によって1993年に作成されました。
# それぞれのカラムは何を表しているのか
# CRIM・・・犯罪発生率（人口単位）
# ZN・・・25,000平方フィート以上の住宅区画の割合
# INDUS・・・非小売業の土地面積の割合（人口単位）
# CHAS・・・チャールズ川沿いかどうか（1:Yes、0:No）
# NOX・・・窒素酸化物の濃度（pphm単位）
# RM・・・1戸あたりの平均部屋数
# AGE・・・1940年よりも前に建てられた家屋の割合
# DIS・・・ボストンの主な5つの雇用圏までの重み付きの郷里
# RAD・・・幹線道路へのアクセス指数
# TAX・・・10,000ドルあたりの所得税率
# PTRATIO・・・教師あたりの生徒の数（人口単位）
# B・・・アフリカ系アメリカ人居住者の割合（人口単位）
# LSTAT・・・低所得者の割合(%)

### 一変数の可視化
X.describe()#全体のカウント、平均、標準偏差、最小値、最大値などを把握
X.info()#欠損値の把握やどのようなカラムがあるかを把握

### describeメソッドを用いることで把握できたこと(考察)を記述せよ
# 前処理に関して気をつけるべき点
# 最大値を見ると、特徴量"B"は396.9に対し、"CHAS"は1だったりする。その為、
# 入力データを正規化してトレーニングしないと精度が出ない恐れがある。
# 機械学習の手法選択に活かせること
# MDEV(住宅価格の中央値)と関連する特徴量が与えられており、教師あり学習の回帰の手法が選択出切ると想定される。
# 機械学習の特徴選択について考えられること
# NOX(窒素酸化物の濃度)よりも、LSTAT(低所得者の割合)やCRIM(犯罪発生率)といった特徴量のほうが
# 住宅価格に影響を与えると考えられる。

### 二変数の可視化　相関
import seaborn as sns
a = pd.concat([X,y],axis=1)
pd.DataFrame.corr(a)
### 二変数の可視化について記述せよ　相関
# そもそも相関とは何か
# 二つ以上の事物の、一方が変われば他方もそれに連れて変わるとか、あるものの影響を
# 受けてかかわり合っているとかいうように、互いに関係を持つこと。また、そういう関係。
# MEDVにもっとも相関があると出力された変数を3つほどあげよ
# LSTAT(-0.737663)、RM(0.695360)、PTRATIO(-0.507787)の3つが挙げられる。
# 相関がある変数を判明させることでどのように利用できそうか考察せよ（調査せよ）
# 学習する際に指定する特徴量とするか否かの判断で利用できる。
### 相関の注意点について記述せよ
# 便利な相関であるが、落とし穴がある
# 相関が正しく2変数の関係を表していない時がある、それはどのような場合か
# 相関係数は、扱うデータの外れ値（スケール）に大きく影響を受けてしまう。
# 外れ値によってデータのスケールが変わってしまう為、SVMやニューラルネットワークの
# 精度も大きく影響を受ける。
# その場合、どのような手段によって、確認するべきか
# 四分位範囲(IQR)等の手法を用いて、はずれ値を確認する。

### 二変数の可視化　プロット
# MEDVにもっとも相関が高かった、変数LSTAT、RM、PTRATIOのプロット図をそれぞれ作成する。
import matplotlib.pyplot as plt
%matplotlib inline
plt.scatter(X['LSTAT'],y)
plt.xlabel('LSTAT')
plt.ylabel('MEDV')
plt.title('MEDV & LSTAT')
plt.show()

plt.scatter(X['RM'],y)
plt.xlabel('RM')
plt.ylabel('MEDV')
plt.title('MEDV & RM')
plt.show()

plt.scatter(X['PTRATIO'],y)
plt.xlabel('PTRATIO')
plt.ylabel('MEDV')
plt.title('MEDV & PTRATIO')
plt.show()

### 二変数の可視化について記述せよ　プロット
# それぞれのプロット結果から、今回の相関はそれぞれ信用できるか記述せよ。
# プロット結果から、LSTATとRMはMDEVと相関があるものの、PTRATIOは相関はなかった。

### データセット分割
from sklearn.model_selection import train_test_split
X_train,X_test,y_train,y_test = train_test_split(X,y,test_size=0.3,random_state=0)

### 学習
from sklearn.tree import DecisionTreeRegressor
regressor1 = DecisionTreeRegressor(max_depth=1)
regressor2 = DecisionTreeRegressor(max_depth=2)
regressor3 = DecisionTreeRegressor(max_depth=3)

regressor1.fit(X_train,y_train)
regressor2.fit(X_train,y_train)
regressor3.fit(X_train,y_train)

### 決定係数
score1 = regressor1.score(X_test,y_test)
score2 = regressor2.score(X_test,y_test)
score3 = regressor3.score(X_test,y_test)
print("score1:",score1)
print("score2:",score2)
print("score3:",score3)

### 学習曲線
import numpy as np
def plot_learning_curve(estimator, title, X, y, ylim=None, cv=None,
                        n_jobs=1, train_sizes=np.linspace(.1, 1.0, 5)):
    """
    Generate a simple plot of the test and training learning curve.

    Parameters
    ----------
    estimator : object type that implements the "fit" and "predict" methods
        An object of that type which is cloned for each validation.

    title : string
        Title for the chart.

    X : array-like, shape (n_samples, n_features)
        Training vector, where n_samples is the number of samples and
        n_features is the number of features.

    y : array-like, shape (n_samples) or (n_samples, n_features), optional
        Target relative to X for classification or regression;
        None for unsupervised learning.

    ylim : tuple, shape (ymin, ymax), optional
        Defines minimum and maximum yvalues plotted.

    cv : int, cross-validation generator or an iterable, optional
        Determines the cross-validation splitting strategy.
        Possible inputs for cv are:
          - None, to use the default 3-fold cross-validation,
          - integer, to specify the number of folds.
          - An object to be used as a cross-validation generator.
          - An iterable yielding train/test splits.

        For integer/None inputs, if ``y`` is binary or multiclass,
        :class:`StratifiedKFold` used. If the estimator is not a classifier
        or if ``y`` is neither binary nor multiclass, :class:`KFold` is used.

        Refer :ref:`User Guide <cross_validation>` for the various
        cross-validators that can be used here.

    n_jobs : integer, optional
        Number of jobs to run in parallel (default 1).
    """
    plt.figure()
    plt.title(title)
    if ylim is not None:
        plt.ylim(*ylim)
    plt.xlabel("Training examples")
    plt.ylabel("Score")
    train_sizes, train_scores, test_scores = learning_curve(
        estimator, X, y, cv=cv, n_jobs=n_jobs, train_sizes=train_sizes)
    train_scores_mean = np.mean(train_scores, axis=1)
    train_scores_std = np.std(train_scores, axis=1)
    test_scores_mean = np.mean(test_scores, axis=1)
    test_scores_std = np.std(test_scores, axis=1)
    plt.grid()

    plt.fill_between(train_sizes, train_scores_mean - train_scores_std,
                     train_scores_mean + train_scores_std, alpha=0.1,
                     color="r")
    plt.fill_between(train_sizes, test_scores_mean - test_scores_std,
                     test_scores_mean + test_scores_std, alpha=0.1, color="g")
    plt.plot(train_sizes, train_scores_mean, 'o-', color="r",
             label="Training score")
    plt.plot(train_sizes, test_scores_mean, 'o-', color="g",
             label="Cross-validation score")

    plt.legend(loc="best")

# メソッド呼び出し　

print(__doc__)
import numpy as np
import matplotlib.pyplot as plt
from sklearn.model_selection import learning_curve
from sklearn.model_selection import ShuffleSplit
from sklearn.tree import DecisionTreeRegressor
title = "Learning Curves (DecisionTreeRegressor)"
# Cross validation with 100 iterations to get smoother mean test and train
# score curves, each time with 20% data randomly selected as a validation set.
cv = ShuffleSplit(n_splits=100, test_size=0.2, random_state=0)

for degree in range(1,11):
    estimator = DecisionTreeRegressor(max_depth=degree)
    plot_learning_curve(estimator, title,X , y, ylim=(0.3, 1.01), cv=cv, n_jobs=4)

### 学習曲線について記述せよ
# High varianceとは何か
# 過学習(overfitting)
# High Biasとは何か
# 未学習(underfitting)
# もっとも結果が良いと思われた深さをひとつ選択せよ
# depth=3(0.587730357252)
# なぜその深さが最も良いか記述せよ
# 過学習が抑えられ、その他の深さと比べて汎用性がある為。
# その最も結果が良いと思われる深さにおいて、データ数を増やしていくと、さらに結果は良くなるか記述せよ
# 学習のサンプルデータが増える為、精度が上がり、結果は良くなる。

### グリッドサーチ
from sklearn.metrics import r2_score, make_scorer#決定係数
from sklearn.model_selection import GridSearchCV

regressor = DecisionTreeRegressor()
params = {"max_depth": list(range(1,11))}
scoring_fnc = make_scorer(r2_score)
grid = GridSearchCV(estimator=regressor, param_grid=params, scoring=scoring_fnc,cv=5)#cs=5は交差検証の回数
grid.fit(X_train,y_train)
print(grid.best_params_)

### グリッドサーチについて記述せよ
# グリッドサーチとは何か
# ハイパーパラメータの組み合わせについて２次元の格子状（網目）のようにしらみつぶしに探す方法
# グリッドサーチと同じようにハイパーパラメーターを探索する手法を挙げよ
# ランダムサーチ
# その手法は、グリッドサーチと比べて、どの点で優れているか、どのように使い分けるべきか
# 組み合わせをランダムに選んで試す為、完了までの時間が短い。
# ディープラーニングなどハイパーパラメータが非常に多い場合、グリッドサーチで
# しらみつぶしにではなく、組み合わせをランダムに選んで試す、ランダムサーチを用いる。

### グリッドサーチを使いこなそう
#グリッドサーチの際、Kfoldの折りたたみの回数をそれぞれ3回と10回に変えて検証
grid__num_3 = GridSearchCV(estimator=regressor, param_grid=params, scoring=scoring_fnc,cv=3)
grid__num_3.fit(X_train,y_train)
print("grid_num_3:",grid__num_3.best_params_)

grid__num_10 = GridSearchCV(estimator=regressor, param_grid=params, scoring=scoring_fnc,cv=10)
grid__num_10.fit(X_train,y_train)
print("grid_num_10:",grid__num_10.best_params_)

#評価方法を決定係数から平均二乗誤差に変更せよ
from sklearn.metrics import mean_squared_error
score = make_scorer(mean_squared_error)
grid_mse = GridSearchCV(estimator=regressor,param_grid=params, scoring=score,cv=5)
grid_mse.fit(X_train,y_train)
print("grid_best_param:",grid_mse.best_params_)

### テスト
x=X.loc[:,['LSTAT','RM',"NOX","AGE","TAX","RAD","CRIM","DIS"]].as_matrix()
X_train,X_test,y_train,y_test = train_test_split(x,y,test_size=0.4,random_state=0)
tree = DecisionTreeRegressor(max_depth=4,random_state=0)
tree.fit(X_train,y_train)
score_tes = tree.score(X_test,y_test)
print("test:",score_tes)
